File: android/app/src/main/java/com/carelog/util/PhiSanitizer.kt
package com.carelog.util

import android.util.Log
import java.util.regex.Pattern

/**
 * PHI (Protected Health Information) sanitizer for logs and crash reports.
 * Ensures HIPAA compliance by removing or masking sensitive data.
 */
object PhiSanitizer {

    private const val TAG = "PhiSanitizer"

    // Patterns for PHI detection
    private val EMAIL_PATTERN = Pattern.compile(
        "[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,}"
    )

    private val PHONE_PATTERN = Pattern.compile(
        "\\+?[0-9]{1,4}?[-.\\s]?\\(?[0-9]{1,3}?\\)?[-.\\s]?[0-9]{1,4}[-.\\s]?[0-9]{1,4}[-.\\s]?[0-9]{1,9}"
    )

    private val SSN_PATTERN = Pattern.compile(
        "\\d{3}[- ]?\\d{2}[- ]?\\d{4}"
    )

    private val DATE_OF_BIRTH_PATTERN = Pattern.compile(
        "(\\d{1,2}[/\\-]\\d{1,2}[/\\-]\\d{2,4})|(\\d{4}[/\\-]\\d{1,2}[/\\-]\\d{1,2})"
    )

    private val UUID_PATTERN = Pattern.compile(
        "[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}"
    )

    // Sensitive field names to redact values for
    private val SENSITIVE_FIELDS = setOf(
        "name", "email", "phone", "address", "ssn", "dob", "dateOfBirth",
        "password", "token", "accessToken", "refreshToken", "authorization",
        "patientName", "doctorName", "attendantName", "relativeName",
        "conditions", "diagnosis", "medication"
    )

    // Replacement strings
    private const val REDACTED = "[REDACTED]"
    private const val MASKED_EMAIL = "[EMAIL]"
    private const val MASKED_PHONE = "[PHONE]"
    private const val MASKED_SSN = "[SSN]"
    private const val MASKED_DOB = "[DOB]"
    private const val MASKED_UUID = "[UUID]"

    /**
     * Sanitize a log message by removing PHI.
     */
    fun sanitize(message: String): String {
        var sanitized = message

        // Mask emails
        sanitized = EMAIL_PATTERN.matcher(sanitized).replaceAll(MASKED_EMAIL)

        // Mask phone numbers
        sanitized = PHONE_PATTERN.matcher(sanitized).replaceAll(MASKED_PHONE)

        // Mask SSNs
        sanitized = SSN_PATTERN.matcher(sanitized).replaceAll(MASKED_SSN)

        // Mask dates that could be DOB
        sanitized = DATE_OF_BIRTH_PATTERN.matcher(sanitized).replaceAll(MASKED_DOB)

        // Mask UUIDs (patient IDs, etc.)
        sanitized = UUID_PATTERN.matcher(sanitized).replaceAll(MASKED_UUID)

        return sanitized
    }

    /**
     * Sanitize a map of key-value pairs.
     */
    fun sanitizeMap(data: Map<String, Any?>): Map<String, Any?> {
        return data.mapValues { (key, value) ->
            if (SENSITIVE_FIELDS.any { key.contains(it, ignoreCase = true) }) {
                REDACTED
            } else {
                when (value) {
                    is String -> sanitize(value)
                    is Map<*, *> -> sanitizeMap(value as Map<String, Any?>)
                    else -> value
                }
            }
        }
    }

    /**
     * Safe logging wrapper that sanitizes before logging.
     */
    fun logDebug(tag: String, message: String) {
        if (BuildConfig.DEBUG) {
            Log.d(tag, sanitize(message))
        }
    }

    fun logInfo(tag: String, message: String) {
        Log.i(tag, sanitize(message))
    }

    fun logWarning(tag: String, message: String) {
        Log.w(tag, sanitize(message))
    }

    fun logError(tag: String, message: String, throwable: Throwable? = null) {
        val sanitizedMessage = sanitize(message)
        val sanitizedThrowable = throwable?.let { sanitizeThrowable(it) }
        Log.e(tag, sanitizedMessage, sanitizedThrowable)
    }

    /**
     * Sanitize exception message and stack trace.
     */
    private fun sanitizeThrowable(throwable: Throwable): Throwable {
        return object : Throwable(sanitize(throwable.message ?: ""), throwable.cause) {
            override fun getStackTrace(): Array<StackTraceElement> {
                return throwable.stackTrace
            }
        }
    }

    /**
     * Create sanitized crash report data.
     */
    fun createSanitizedCrashData(
        throwable: Throwable,
        additionalData: Map<String, Any?> = emptyMap()
    ): Map<String, Any?> {
        return mapOf(
            "error_type" to throwable::class.java.simpleName,
            "error_message" to sanitize(throwable.message ?: "Unknown error"),
            "stack_trace" to throwable.stackTrace.take(10).map { it.toString() },
            "additional_data" to sanitizeMap(additionalData)
        )
    }
}

/**
 * BuildConfig placeholder for release checks.
 */
private object BuildConfig {
    val DEBUG = false // Would be set by actual build config
}

File: ios/CareLog/CareLog/Util/PhiSanitizer.swift
import Foundation
import os.log

/// PHI (Protected Health Information) sanitizer for logs and crash reports.
/// Ensures HIPAA compliance by removing or masking sensitive data.
enum PhiSanitizer {

    // MARK: - Patterns

    private static let emailPattern = try! NSRegularExpression(
        pattern: "[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\\.[a-zA-Z]{2,}",
        options: []
    )

    private static let phonePattern = try! NSRegularExpression(
        pattern: "\\+?[0-9]{1,4}?[-.\\s]?\\(?[0-9]{1,3}?\\)?[-.\\s]?[0-9]{1,4}[-.\\s]?[0-9]{1,4}[-.\\s]?[0-9]{1,9}",
        options: []
    )

    private static let ssnPattern = try! NSRegularExpression(
        pattern: "\\d{3}[- ]?\\d{2}[- ]?\\d{4}",
        options: []
    )

    private static let dobPattern = try! NSRegularExpression(
        pattern: "(\\d{1,2}[/\\-]\\d{1,2}[/\\-]\\d{2,4})|(\\d{4}[/\\-]\\d{1,2}[/\\-]\\d{1,2})",
        options: []
    )

    private static let uuidPattern = try! NSRegularExpression(
        pattern: "[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}",
        options: .caseInsensitive
    )

    // MARK: - Sensitive Fields

    private static let sensitiveFields: Set<String> = [
        "name", "email", "phone", "address", "ssn", "dob", "dateOfBirth",
        "password", "token", "accessToken", "refreshToken", "authorization",
        "patientName", "doctorName", "attendantName", "relativeName",
        "conditions", "diagnosis", "medication"
    ]

    // MARK: - Replacement Strings

    private static let redacted = "[REDACTED]"
    private static let maskedEmail = "[EMAIL]"
    private static let maskedPhone = "[PHONE]"
    private static let maskedSsn = "[SSN]"
    private static let maskedDob = "[DOB]"
    private static let maskedUuid = "[UUID]"

    // MARK: - Sanitization

    /// Sanitize a string by removing PHI.
    static func sanitize(_ message: String) -> String {
        var sanitized = message
        let range = NSRange(sanitized.startIndex..., in: sanitized)

        // Mask emails
        sanitized = emailPattern.stringByReplacingMatches(
            in: sanitized,
            range: range,
            withTemplate: maskedEmail
        )

        // Mask phone numbers
        let phoneRange = NSRange(sanitized.startIndex..., in: sanitized)
        sanitized = phonePattern.stringByReplacingMatches(
            in: sanitized,
            range: phoneRange,
            withTemplate: maskedPhone
        )

        // Mask SSNs
        let ssnRange = NSRange(sanitized.startIndex..., in: sanitized)
        sanitized = ssnPattern.stringByReplacingMatches(
            in: sanitized,
            range: ssnRange,
            withTemplate: maskedSsn
        )

        // Mask dates
        let dobRange = NSRange(sanitized.startIndex..., in: sanitized)
        sanitized = dobPattern.stringByReplacingMatches(
            in: sanitized,
            range: dobRange,
            withTemplate: maskedDob
        )

        // Mask UUIDs
        let uuidRange = NSRange(sanitized.startIndex..., in: sanitized)
        sanitized = uuidPattern.stringByReplacingMatches(
            in: sanitized,
            range: uuidRange,
            withTemplate: maskedUuid
        )

        return sanitized
    }

    /// Sanitize a dictionary by removing PHI from values.
    static func sanitize(_ dictionary: [String: Any]) -> [String: Any] {
        var sanitized = [String: Any]()

        for (key, value) in dictionary {
            if sensitiveFields.contains(where: { key.localizedCaseInsensitiveContains($0) }) {
                sanitized[key] = redacted
            } else if let stringValue = value as? String {
                sanitized[key] = sanitize(stringValue)
            } else if let dictValue = value as? [String: Any] {
                sanitized[key] = sanitize(dictValue)
            } else {
                sanitized[key] = value
            }
        }

        return sanitized
    }

    // MARK: - Safe Logging

    private static let logger = Logger(subsystem: "com.carelog", category: "Sanitized")

    /// Log debug message with PHI sanitization.
    static func logDebug(_ message: String) {
        #if DEBUG
        logger.debug("\(sanitize(message), privacy: .public)")
        #endif
    }

    /// Log info message with PHI sanitization.
    static func logInfo(_ message: String) {
        logger.info("\(sanitize(message), privacy: .public)")
    }

    /// Log warning message with PHI sanitization.
    static func logWarning(_ message: String) {
        logger.warning("\(sanitize(message), privacy: .public)")
    }

    /// Log error message with PHI sanitization.
    static func logError(_ message: String, error: Error? = nil) {
        let sanitizedMessage = sanitize(message)
        if let error = error {
            let sanitizedError = sanitize(error.localizedDescription)
            logger.error("\(sanitizedMessage, privacy: .public): \(sanitizedError, privacy: .public)")
        } else {
            logger.error("\(sanitizedMessage, privacy: .public)")
        }
    }

    // MARK: - Crash Report Data

    /// Create sanitized crash report data.
    static func createCrashData(
        error: Error,
        additionalData: [String: Any] = [:]
    ) -> [String: Any] {
        let nsError = error as NSError

        return [
            "error_domain": nsError.domain,
            "error_code": nsError.code,
            "error_message": sanitize(nsError.localizedDescription),
            "additional_data": sanitize(additionalData)
        ]
    }
}

// MARK: - String Extension

extension String {
    /// Returns a PHI-sanitized version of the string.
    var sanitized: String {
        PhiSanitizer.sanitize(self)
    }
}
